### PR TITLE
Geometry in integrators

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp [hunk 1]
```diff
@@ -5,11 +5,14 @@
 
 #include "base/not_null.hpp"
 #include "integrators/symplectic_integrator.hpp"
+#include "quantities/named_quantities.hpp"
 
 namespace principia {
 
 using base::not_null;
+using quantities::Difference;
```
Comment (pleroy) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Shouldn't these new `using` go into the body?

File: mathematica/integrator_plots.cpp [hunk 2]
```diff
@@ -262,6 +294,172 @@ void GenerateKeplerProblemWorkErrorGraphs() {
...
+  // We use dense sampling in order to compute average errors, this leads to
+  // more evaluations than reported for FSAL methods.
+  parameters.sampling_period = 1;
+
+  SRKNIntegrator::Solution<Position<ICRFJ2000Ecliptic>,
+                           Velocity<ICRFJ2000Ecliptic>> reference_solution;
+  {
+    std::vector<
+      SRKNIntegrator::Solution<Position<ICRFJ2000Ecliptic>,
+      Velocity<ICRFJ2000Ecliptic>>> reference_solutions;
```
Comment (pleroy) [inline mathematica/integrator_plots.cpp]: Confusing indent.

File: mathematica/integrator_plots.cpp [hunk 3]
```diff
@@ -262,6 +294,172 @@ void GenerateKeplerProblemWorkErrorGraphs() {
...
+      method.integrator->
+          SolveTrivialKineticEnergyIncrement<Position<ICRFJ2000Ecliptic>>(
+              std::bind(ComputeGravitationalAcceleration<ICRFJ2000Ecliptic>,
+                        _1, _2, _3, std::cref(bodies)),
+              parameters,
+              &reference_solutions.back());
+    }
+    int const reference_size = reference_solutions.front().size();
+    for (auto const& solution : reference_solutions) {
+      CHECK(solution.size() == reference_size);
```
Comment (pleroy) [inline mathematica/integrator_plots.cpp]: `CHECK_EQ`?

File: quantities/elementary_functions_body.hpp [hunk 4]
```diff
@@ -1,26 +1,26 @@
...
-                   Q::Dimensions::Time & 1 ||
-                   Q::Dimensions::Current & 1 ||
-                   Q::Dimensions::Temperature & 1 ||
-                   Q::Dimensions::Amount & 1 ||
-                   Q::Dimensions::LuminousIntensity & 1 ||
-                   Q::Dimensions::Winding & 1 ||
-                   Q::Dimensions::Angle & 1 ||
-                   Q::Dimensions::SolidAngle & 1)>> {
+struct SquareRootGenerator<Q, std::enable_if_t<!(Q::Dimensions::Length & 1 ||
+                                     Q::Dimensions::Mass & 1 ||
```
Comment (pleroy) [inline quantities/elementary_functions_body.hpp]: Bad indent, align all the `Q`s.

File: quantities/quantities.hpp [hunk 5]
```diff
@@ -149,11 +168,13 @@ class Quantity {
   double magnitude_;
 
   template<typename LDimensions, typename RDimensions>
-  friend Product<Quantity<LDimensions>, Quantity<RDimensions>> operator*(
+  friend internal::Product<Quantity<LDimensions>, Quantity<RDimensions>>
+  operator*(
       Quantity<LDimensions> const& left,
```
Comment (pleroy) [inline quantities/quantities.hpp]: Would fit on the previous line.

File: quantities/quantities.hpp [hunk 6]
```diff
@@ -149,11 +168,13 @@ class Quantity {
...
       Quantity<RDimensions> const& right);
   template<typename LDimensions, typename RDimensions>
-  friend Quotient<Quantity<LDimensions>, Quantity<RDimensions>> operator/(
+  friend internal::Quotient<Quantity<LDimensions>, Quantity<RDimensions>>
+  operator/(
       Quantity<LDimensions> const& left,
```
Comment (pleroy) [inline quantities/quantities.hpp]: Same.

File: testing_utilities/numerical_analysis_body.hpp [hunk 7]
```diff
@@ -8,7 +8,10 @@
 
 namespace principia {
 
+using geometry::Displacement;
+using geometry::InnerProduct;
 using quantities::Force;
+using quantities::Exponentiation;
```
Comment (pleroy) [inline testing_utilities/numerical_analysis_body.hpp]: Alphabetize.

**Comment by pleroy** [comment]
Please run the benchmarks before and after to make sure that this doesn't affect performance.

File: testing_utilities/numerical_analysis_body.hpp (hunk 7 above)
Comment (eggrobin) [inline testing_utilities/numerical_analysis_body.hpp]: Deno.

File: quantities/quantities.hpp (hunk 6 above)
Comment (eggrobin) [inline quantities/quantities.hpp]: Done.

File: quantities/quantities.hpp (hunk 5 above)
Comment (eggrobin) [inline quantities/quantities.hpp]: Done.

File: quantities/elementary_functions_body.hpp (hunk 4 above)
Comment (eggrobin) [inline quantities/elementary_functions_body.hpp]: Done.

File: mathematica/integrator_plots.cpp (hunk 3 above)
Comment (eggrobin) [inline mathematica/integrator_plots.cpp]: Done.

File: mathematica/integrator_plots.cpp (hunk 2 above)
Comment (eggrobin) [inline mathematica/integrator_plots.cpp]: Done.

File: integrators/symplectic_runge_kutta_nystrom_integrator.hpp (hunk 1 above)
Comment (eggrobin) [inline integrators/symplectic_runge_kutta_nystrom_integrator.hpp]: Except `Variation`, yes.

**Comment by eggrobin** [comment]
Benchmark results, μ and σ only, only a few methods, for readability:
After:
```
[geometry-in-integrators]> .\Release\benchmarks.exe --benchmark_repetitions=5 --benchmark_min_time=10 --benchmark_filter=HarmonicOscillator
Benchmarking on 1 X 2672 MHz CPU
2015/04/17-12:00:33
Benchmark                                                                                                           Time(ns)    CPU(ns) Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Leapfrog>_mean                                            3289710775 3279141020          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Leapfrog>_stddev                                            57962013   60338315          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::PseudoLeapfrog>_mean                                      3334768215 3294741120          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::PseudoLeapfrog>_stddev                                      64231304   26839427          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::McLachlanAtela1992Order2Optimal>_mean                     3416491571 3347781460          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::McLachlanAtela1992Order2Optimal>_stddev                     94937798   52764326          5
BM_SolveHarmonicOscillator<SRKNIntegrator,&integrators::McLachlanAtela1992Order5Optimal>_mean                     4521039453 4474108680          5
BM_SolveHarmonicOscillator<SRKNIntegrator,&integrators::McLachlanAtela1992Order5Optimal>_stddev                     56879278   26839427          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Yoshida1990Order8D>_mean                                  7486808554 7453727780          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Yoshida1990Order8D>_stddev                                  13495694   26839427          5
```
Before:
```
[(ad2ba27...)]> .\Release\benchmarks.exe --benchmark_repetitions=5 --benchmark_min_time=10 --benchmark_filter=HarmonicOscillator
Benchmarking on 1 X 2672 MHz CPU
2015/04/17-12:13:40
Benchmark                                                                                                           Time(ns)    CPU(ns) Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Leapfrog>_mean                                            3445049921 3385221700          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Leapfrog>_stddev                                            72000239   51267160          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::PseudoLeapfrog>_mean                                      3858345524 3541222700          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::PseudoLeapfrog>_stddev                                     226402760  153167244          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::McLachlanAtela1992Order2Optimal>_mean                     3689404789 3544342720          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::McLachlanAtela1992Order2Optimal>_stddev                    188941173  171004422          5
BM_SolveHarmonicOscillator<SRKNIntegrator,&integrators::McLachlanAtela1992Order5Optimal>_mean                     4471437517 4458508580          5
BM_SolveHarmonicOscillator<SRKNIntegrator,&integrators::McLachlanAtela1992Order5Optimal>_stddev                     18968217   11674046          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Yoshida1990Order8D>_mean                                  7716337791 7687729280          5
BM_SolveHarmonicOscillator<SPRKIntegrator,&integrators::Yoshida1990Order8D>_stddev                                  37523154   30249716          5

```

**Pull request merged**